### DISCUSSION TITLE
Research project on Sustainable Computing

### DISCUSSION THREAD

**Discussion by seanmcilroy29** [Green-Software-Foundation/sci#348] [General]
From Chris Xie (Huawei) 
We are intending to initiate a research project on Sustainable Computing topic, in particular, how programming languages (Rust, C, Java, Python etc.) affect the carbon efficiencies of the full computing stack, and translate the efficiencies into economic impacts in dollar terms, using the SCI methodology as a tool for such studies. Effectively this project could be a case study or an application of SCI and intends to apply the SCI methodology to the Social Cost of Carbon (SCC) study of programming languages and software systems. Apply meaning to SCI scores is not limited to dollar terms, but it could also be carbon efficiency rankings of comparable computing systems, ICT solutions, 5G base stations, etc. Please see the attached document for details. Please feel free to ping me for further discussions, ideas, collaborations. - [SoftwareSCC_GSF.pdf](https://github.com/Green-Software-Foundation/standards-wg/files/10512336/SoftwareSCC_GSF.pdf)

**Comment by Henry-WattTime** [comment]
Henry: good luck identifying a social cost of carbon
Tammy: efficiency of software isn't the primary driver, it is how the underlying hardware interprets the software
w/ Srini studied different languages doing similar tasks. Is is software? Hardware? How to quantify? 
Ben: If interpreter is the same, shouldn't the software be the same? What is the scale of the difference? Marginal gains or significant? Lots of existing research around this.
New research would be as it relates to SCI
GSF position? Does it matter or not matter? It's a tool in the toolbelt, so it could have an effect?
To switch, need a reason/ROI. Value/social cost could be a way to tie it back to value. 
Gadhu: everything else remaining the same, how much does changing the language have an effect, what's the scale? Is this worth investing in?
Cost of the work vs overall gain in total emissions reduction
value could be combined cost and overall value.
(henry example, lots of little apps vs one large app
Srini: Performance efficiency discussion tend to be theoretical. Study comparison of languages.
Motivation: Choices related to performance requirements and key requirements. Don't optimize unless required to. Emissions may be a new incentive to care about performance. 
Good for greenfield development, emissions could be a factor
Inherent efficiency of language, and quality of the actual code.

Question for Asim? Python is ~60% less efficient than java, implies switching, but could fix Python
Marcos: can we work with the open source community to increase the efficiency of software? Open area to explore
-when in the thick of development, can forget efficiency, efficiency in the language side helps

Navveen: wrote application in nodejs and rust, keep stuff constant, rust was very efficient (~5-10%). 
hardware/users/etc the same
tradeoffs, limited libraries and APIs, new language
consider tradeoffs
case study still internal, calculated SCI score

RESOURCES:
Rust comparison: https://aws.amazon.com/blogs/opensource/sustainability-with-rust/
Energy efficiency of Languages: https://greenlab.di.uminho.pt/wp-content/uploads/2017/09/paperSLE.pdf
Ziliang at Texas State has done some research into this
Social Cost of Carbon: https://www.rff.org/publications/journal-articles/comprehensive-evidence-implies-a-higher-social-cost-of-co2/

Core question:
Same application with SCI score, ported to different language, what is impact on score (operational/energy emissions only, no change in hardware and embodied emissions)?

**Comment by Henry-WattTime** [comment]
WG: 
Navveen: We had discussed creating a reference application earlier for SCI and comparing languages, with other factors being the same. Trade-offs on computing, skills, and expertise can provide cost angle and simulation. Benchmarks would be useful in this exercise
Ciril: +1 with Navveen. There seem to be two steps: 1. Comparing Languages 2. Assessing Pricing of the associated Carbon

Gadhu: What will be really useful is just some detailed benchmarking of various aspects of the stack  - language, pattern etc for achieving a particular function so that you can compare the carbon cost of using these aspects to design your software system.
Help enable a decision about what language to use. When making a carbon decision, low level aspects. Create benchmarks for different architectures. SCI will depend on many factors, so in this case keep everything but the language the same between comparison (same hardware, location, time). Then you can know which language is more carbon efficient. Then you can optimize other parameters (hardware, switches, etc). This would allow a baseline comparison. Unlikely that this research would be done in a corporate setting, good for academic study. See Ziliang' study that compares ML models does this for model comparisons.

Pindy: comparing languages - could we do a longitudinal study compare the same solutions using different languages - can the professor create simulations

**Comment by Henry-WattTime** [comment]
update in August

**Comment by chrisxie-fw** [comment]
During today's discussion (5/25/2023), I shared an idea and an ultimate vision that someday, major github open source projects will have an SCI score or SSC (Social Cost of Carbon dollars sign) along side the number of forks and likes, or some type of carbon efficiency rating of the software, in absolute terms or relative terms, or some sort of carbon efficiency certification. To achieve that, one of the approaches is to define a standard/specification of a controlled environment (hardware, software, versions of SCI spec, locale etc)  where such rating/evaluation of SCI/SCC is produced. This way, anyone can take such an environmental specification and produce an SCI/SSC dollars of their own software systems. The produced SCI/SCC scores will be compliant and certified by GSF. Of course, it will make sense for GSF as an organization to build an initial reference implementation of the controlled environment, and ideally the compliant environment built by 3rd parties should produce similar SCI/SCC scores results. This could also be an extension of the sustainable computing research project. This idea is a bit like UL Labs that measures product safety. In this cause, software product or systems are measured for its carbon footprint, in CO2's or in SSC dollars. The point of this idea is to translate SCI scores to real life meaning, real life value, in terms of dollar amount, or understandable certification or meaning of carbon efficiency of software systems. Again, the initial scope of the standard is to define the parameters of the controlled testing environment in order to create the SCI/SCC scores, so that anyone can take the standard specification to build their own testing lab environment to test their software to create their own SCI/SCC score, and GSF could issue the certificate that their score meets or complies with GSF's standard of certification. GSF still maintains the authority of the certificate and the evolution of the SCI/SCC spec, and still is responsible for the definition of the SCI, SCC spec, the testing environment spec, and the future iterations of these specifications. If this vision is achieved, SCI will have a really meaningful impact to the whole software industry. The value of SCI standard will be truly realized. Look forward to continued brainstorming of ideas of translating SCI to real life meanings and applications.

**Comment by Henry-WattTime** [comment]
WG: recommends merging with SCER

**Comment by ArneTR** [comment]
Hey @chrisxie-fw , 

we have just submitted an article that describes an automated open source approach to measure the SCI. Is that something that fits what you are suggesting?

https://github.com/Green-Software-Foundation/sci-guide/issues/52

@Henry-WattTime 
We also saw the project from the GSF to have the SCER for many applications. We have done something similar in the past by using for instance TPC Benchmarks for databases as standardized usage scenario and then running automated SCI benchmarks against it.

See an example for a TPC Select benchmark here for:
- PostgreSQL: https://metrics.green-coding.berlin/stats.html?id=69958765-14bb-44c0-90cd-8ef084c0ecdd
- MariaDB: https://metrics.green-coding.berlin/stats.html?id=c818a440-15b8-4da9-9614-96c3296fbb8a
- Comparison-View: https://metrics.green-coding.berlin/compare.html?ids=c818a440-15b8-4da9-9614-96c3296fbb8a,69958765-14bb-44c0-90cd-8ef084c0ecdd
 
<img width="539" alt="Screenshot 2023-09-20 at 12 05 22 PM" src="https://github.com/Green-Software-Foundation/sci/assets/250671/4f5a89e9-2cfb-4c37-b062-9a17a380480e">

The repo to match it is here: https://github.com/green-coding-berlin/DBMS-bench

Is that something that could be interesting for the SCER project?

**Reply by Henry-WattTime** [reply]
Hello @ArneTR, the TPC benchmarks are very interesting and very applicable to the SCER conversation. @mvaltas, is this something to look at in more detail for the SCER?

**Reply by ghost** [reply]
Yes, definitelly it is something to look at, the result of mgCO2E per TPC-C select looks very well within the bounds of what want to grade software in.

**Reply by ArneTR** [reply]
Just for cross-referencing: I also made a post in the SCER project about other possible benchmarks on this topic like for browsers etc. : https://github.com/Green-Software-Foundation/scer/issues/1

**Comment by chrisxie-fw** [comment]
Hi @ArneTR ,  you presented a great tool that could potentially be helpful in SCER project, thank you! 

I was thinking if some kind of labelling be possible for different software projects, similar to what France has done to food nutritional score labelling: [Nutri-Score](https://article.innovamarketinsights360.com/articleimgs/article_images/637516695831662303nutriscore%20FOPL%20label%20front%20of%20pack%20nutrition.jpg): https://en.wikipedia.org/wiki/Nutri-Score . 
![637516695831662303nutriscore FOPL label front of pack nutrition](https://github.com/Green-Software-Foundation/sci/assets/54118227/5c92be8c-3394-42b9-a439-2495a59c2872) 

- Referencing nutri-score's algorithm, what would be the key factors that determine the carbon efficiency of a certain piece of software, noting that SCI score is an absolute value, and label rating is a relative value.
- In the case of PostgreSQL and MariaDB mentioned above, in that test scenario, for example, if PostgreSQL got a SCI score of 1, MariaDB got a 2, can we then label PostgresSQL an SCER label of "A", MariaDB an SCER label of "B"? Again, these are relative labelling, in its particular test cases. Would the labelling be useful? to who? in what circumstances? 
- Another example: does it make sense to measure SCI scores of idle laptops, one with windows OS, one with Linux OS, one with macOS, of similar hardware configurations? This may relate to the application boundaries of SCI, as in: what areas of applications that SCI score is most meaningful and makes the most sense. SCER may be an extension of SCI that presents SCI in relative terms, like the Nutri-Score.

**Reply by ArneTR** [reply]
The labels is quite interesting, as it provides some visual cues and is already known.

The problem I see with a label like this is that for me as a consumer it looks like something "absolute". For instance here in the EU we have an energy efficiency index for electronics and such. This was recently "updated" meaning that every A++ fridge is now only a C fridge, since energy efficiency overall has gotten better.

For consumers this is extremely confusing. If you apply the same logic to an even faster moving field like software you would have to re-baseline the label at certain intervals.

What we personally find more stable (though slightly less catchy) is the approach that also the WebsiteCarbon site is taking. (See screenshot)
<img width="407" alt="Screenshot 2023-10-16 at 11 12 36 AM" src="https://github.com/Green-Software-Foundation/sci/assets/250671/5cd900c5-0323-4d58-8dce-3ed6bd7e00ce">

The idea is to bind the rating to the percentile where your software is in. One could then also include the A/B/C/D/E/F categories, but it is clearer that the label is more dynamic.

Once we have enough data for a software (say 20+ which is often a ballpark threshold in frequentist statistics) such a rating could be applied.

What do you think?

On a different note: Is there any projects you are running internally at Huawei on this topic?

**Reply by chrisxie-fw** [reply]
I really like the "Cleaner than 90% of pages tested", because it gives a relative rating of the page/website in comparison to the others tested. The relative rating scheme is what the rating of "A to D/F" is meant to convey to the end user.

**Reply by chrisxie-fw** [reply]
One other thing I would like to propose is regarding "0.1g of CO2/view". 0.1g of CO2 may not ring a bell for an average visitor of the website, but if it is converted to [dollars in Social Cost of Carbon](https://www.brookings.edu/articles/what-is-the-social-cost-of-carbon/#:~:text=The%20social%20cost%20of%20carbon%20(SCC)%20is%20an%20estimate%20of,a%20ton%20of%20carbon%20emissions.), that may start to get more attentions to the carbon rating of the site. The same logic may go with the software carbon efficiency rating SCER.

**Reply by chrisxie-fw** [reply]
One last thing is with the scope of the SCER standard. Our thinking is that the Standard may be more about generic framework of getting to a relative carbon efficiency rating, rather than the application specific rating, because there are just too many software applications to compute the rating. So our thinking is to extract the common process out of the calculating the rating, and make that a standard, so that each application specific domain can use this standard framework to create the carbon rating of  their own application domain, such as database software, operating system software, ML software, etc. So we may want to identify the common process of getting a CE (Carbon Efficiency) rating, which may involve the definition of a benchmark (e.g. the most commonly used operations/functions of the software, in the case of database, it may be read/write operations), and the relative CE rating calculation formula or algorithm (in the case of database, the SCI scores of several databases from different vendors are submitted, the algorithm may be to calculate the average rating of each database, and assign a label of A to D based on their relative position in the submitted SCI score range). The key principles here are (1) the standard is generic, common, non-application-specific, (2) relativity in the labelling/rating.

**Reply by ArneTR** [reply]
I really like the Social Cost of Carbon! Interestingly seems to me that it varies by country ... I would have expected it to be a world-wide number, since this is a world-wide problem ... maybe too easy thinking :)

I will bring this up with our team internally and we will probalby bring it into the next release of our tool next week.

What we currently have is this: 
<img width="2266" alt="Screenshot 2023-10-17 at 8 20 24 AM" src="https://github.com/Green-Software-Foundation/sci/assets/250671/8a5108c5-8623-4a03-8d29-6b504f6433bf">

This was the typical form the EPA calculators and such displayed the CO2e data in a more tangible format. The SCC is way better IMHO.

2. I also do agree on the pre-defined standard benchmarks that software has to adhere rather the other way round.
Especially for the very different world of E-Mail programs one can define 2-3 very basic SCER benchmarks where a software has to hit certain "checkpoints / endpoints" and then you get a comparable rating.
Obviously the software has much more but this is then the "luxury" part and up to the user to wage if he wants that for the maybe increased runtime cost.

A place where this has worked well I think is for instance with cars, where a "standard" test is run how much liter per kilometer a car consumes.
Or in graphic-cards and drivers. Stuff like CineBench and 3DMark have just proposed what they thought is reasonable (to my knowledge at least) and created an industry standard benchmark around this.

I think what would make this tangible for the user is to create an examplary case on an application that has so far no standard tests around. DBs have TPC-C already around .... So I guess doing this for an E-Mail application or a calender or a word processor would be very interesting to illustrate the concept and get the community working on drafting and deciding for standard tests.

What do you think? Is that maybe even something the WG is already planning?

**Reply by chrisxie-fw** [reply]
The "Trees, Gasoline..." tangible depictions look great! If SCC is incorporated, that would be fantastic!

I envision that SCER standard may help create a standard framework, or a "meta" standard (standard of standards), that helps create more application specific standards like "CineBench", but in Carbon terms, or SCC terms.
